### ISSUE TITLE
TODOs

### ISSUE THREAD

**Issue by fxmarty** [fxmarty/accelerated-pytorch-transformers-generation#3]
- [x] Experiment with `torch.compile` (find torchdynamo graph breaks and remove them if possible?) 
- [ ] Experiment with iteratively allocated KV cache, see [this suggestion](https://github.com/huggingface/text-generation-inference/issues/376), this would avoid aten::slice calls
- [ ] Can we avoid aten::copy_ calls and aten::slice calls?
- [ ] Test on CPU
- [ ] Support cross-attention
- [ ] Support encoder-decoder architectures
- [x] Support preallocated `attention_mask`
- [ ] Support preallocated `token_type_ids` (llama does not use it though)
- [x] Would a single `qkv_proj` help instead of the current `q_proj` separate from `kv_proj`? [yes, helps slightly]
- [ ] Make sure the implementation is valid not on the argmax but on the logits directly
- [ ] Investigate whether stopping to inherit from nn.Module may help, see [this `__getattr__` overhead](https://github.com/fxmarty/accelerated-pytorch-transformers-generation/assets/9808326/8d5c4fc3-d8c6-4d82-a466-78edf2fa9617)
